### PR TITLE
UX Enhancement for Middleware Override Workflow

### LINKED ISSUE #8335: UX Enhancement for Middleware Override Workflow.
**Is your feature request related to a problem? Please describe.**

Currently, when the middleware of an asset is taken from the facility configuration page or location management page, the middleware field is auto-filled as a placeholder. By hovering over the exclamation button, the user can understand from which page this middleware is pre-filled.

However, new users are raising concerns that they are finding it difficult to understand the priority-based asset middleware overrides. For example, the middleware specified in the asset configuration is given the highest priority, followed by location management, and lastly, the facility configuration 
 
image
image

**Describe the solution you'd like**

Screenshot 2024-08-16 161325

- [ ] When the asset configuration is pre-filled from the location management or facility page, instead of making the user hover over the exclamation mark, add fine text below the input box, along with the pre-filled middleware placeholder, to provide details about the middleware's source.

image

- [ ] Implement a similar pre-filling behavior in the placeholder and fine text for the location management edit page.
- [ ] In the location creation form, under the middleware field, add fine text indicating that if left blank, the facility middleware will be applied to assets.

image

- [ ] In the location management card, display the facility configuration middleware with a "Pre-filled Facility Middleware" badge if no middleware is set for the location.

image

- [ ] Under the facility configuration middleware field, add fine text stating that the facility middleware will be applied to all assets when asset and location middleware are Unspecified.

Note: 

- In the relevant places, Try to add the facility name or location in the fine text so users can easily understand the exact source, rather than a general fine text message.
- The attached UI is a sample mock-up for reference. Feel free to share the proper UI mock-up on the issue before starting work on it.

### PR DESCRIPTION
**Pull request by saikiranpatil**
## Proposed Changes

- Closes #8335 
- Changes
    image
    - [x] When the asset configuration is pre-filled from the location management or facility page, instead of making the user hover over the exclamation mark, add fine text below the input box, along with the pre-filled middleware placeholder, to provide details about the middleware's source.
    
    image
    - [x] In the location creation form, under the middleware field, add fine text indicating that if left blank, the facility middleware will be applied to assets.

    image
    - [x] In the location management card, display the facility configuration middleware with a "Pre-filled Facility Middleware" badge if no middleware is set for the location.

    
    image
    - [x] Under the facility configuration middleware field, add fine text stating that the facility middleware will be applied to all assets when asset and location middleware are Unspecified.

@ohcnetwork/care-fe-code-reviewers

## Merge Checklist

- [ ] Add specs that demonstrate bug / test a new feature.
- [ ] Update [product documentation](https://docs.ohc.network).
- [ ] Ensure that UI text is kept in I18n files.
- [ ] Prep screenshot or demo video for changelog entry, and attach it to issue.
- [ ] Request for Peer Reviews
- [ ] Completion of QA

## Summary by CodeRabbit

- **New Features**
	- Introduced new message properties in various form fields for improved user guidance.
	- Enhanced the `LocationManagement` component to dynamically display middleware information with badges.
	- Refactored form handling in `FacilityConfigure` using `react-hook-form` and Zod for validation.
	- Added new components for improved form structure and accessibility, including `FormField`, `FormItem`, and related components.
	- Added a new `CancelButton` component for improved user interaction.
	- Expanded `AuthorizedButton` component to include an optional `variant` property for additional styling.

- **Bug Fixes**
	- Streamlined error handling and display logic in several components.

- **Documentation**
	- Updated documentation to reflect changes in form field properties and new components.

- **Refactor**
	- Refactored multiple components to improve structure and maintainability, including the introduction of context for form fields.

- **Style**
	- Improved UI elements, including label rendering and error message displays.

### CODE REVIEW

**Comment by saikiranpatil** [comment]
@Jacobjeevan @rithviknishad i would like to clarify. with lot of efforts i would make commits and ask for review several times only to have no response most of the time. i had closed before issues for that reason.

**Comment by Jacobjeevan** [comment]
> @Jacobjeevan @rithviknishad i would like to clarify. with lot of efforts i would make commits and ask for review several times only to have no response most of the time. i had closed before issues for that reason.

We miss a few things with our own work as well as reviewing all the other PRs, but feel free to ping us again on Github or on Slack if we haven't gotten back to you.

(For future, try to maintain within one PR - the other PR was still in a workable state).

**Comment by saikiranpatil** [comment]
> We miss a few things with our own work as well as reviewing all the other PRs, but feel free to ping us again on Github or on Slack if we haven't gotten back to you.
> 
> (For future, try to maintain within one PR - the other PR was still in a workable state).

i do not have access to that PR. will take care in future.
please review the changes required in this PR

File: src/components/Facility/FacilityConfigure.tsx [hunk 1]
```diff
@@ -1,112 +1,74 @@
...
-      return {
-        ...state,
-        errors: action.errors,
-      };
-    }
-    default:
-      return state;
-  }
-};
+import { Submit } from "../Common/ButtonV2";
```
Comment (Jacobjeevan) [inline src/components/Facility/FacilityConfigure.tsx]: Minor thing, but switch to using Button from src/components/ui here as well as in LocationManagement.

File: src/components/Facility/FacilityConfigure.tsx [hunk 2]
```diff
@@ -1,112 +1,74 @@
...
+      ward: data.ward,
+      middleware_address: values.middleware_address,
     };
 
+    setIsLoading(true);
+
     const { res, error } = await request(routes.partialUpdateFacility, {
       pathParams: { id: facilityId },
-      body: data,
+      body: formData,
```
Comment (Jacobjeevan) [inline src/components/Facility/FacilityConfigure.tsx]: Go ahead and switch this to using `useMutate` (similar to useQuery above).

**Review by Jacobjeevan — CHANGES_REQUESTED** [review]
Other the minor changes, lgtm.

**Comment by saikiranpatil** [comment]
@ohcnetwork/care-fe-code-reviewers check the recent changes

File: src/components/Facility/FacilityConfigure.tsx (hunk 2 above)
Comment (saikiranpatil) [reply src/components/Facility/FacilityConfigure.tsx]: done

File: src/components/Facility/FacilityConfigure.tsx (hunk 1 above)
Comment (saikiranpatil) [reply src/components/Facility/FacilityConfigure.tsx]: done

File: src/components/Facility/LocationManagement.tsx [hunk 3]
```diff
@@ -1,9 +1,12 @@
+import { useQuery } from "@tanstack/react-query";
 import { useState } from "react";
 
 import RecordMeta from "@/CAREUI/display/RecordMeta";
 import CareIcon from "@/CAREUI/icons/CareIcon";
 import PaginatedList from "@/CAREUI/misc/PaginatedList";
 
+import { Badge } from "@/components/ui/badge";
+
 import ButtonV2, { Cancel } from "@/components/Common/ButtonV2";
```
Comment (Jacobjeevan) [inline src/components/Facility/LocationManagement.tsx]: As mentioned earlier, Button here as well 👍

File: src/components/Facility/FacilityConfigure.tsx [hunk 4]
```diff
@@ -1,166 +1,147 @@
...
-      return {
-        ...state,
-        errors: action.errors,
-      };
-    }
-    default:
-      return state;
-  }
-};
+import { Button } from "../ui/button";
```
Comment (Jacobjeevan) [inline src/components/Facility/FacilityConfigure.tsx]: Switch to absolute imports

**Review by Jacobjeevan — COMMENTED** [review]
Other than that, lgtm.

Do the below and take care of merge conflict 👍

File: src/components/Facility/LocationManagement.tsx (hunk 3 above)
Comment (saikiranpatil) [reply src/components/Facility/LocationManagement.tsx]: its shadcn badge. do you wish to make it from shadcn button?

File: src/components/Facility/FacilityConfigure.tsx (hunk 4 above)
Comment (saikiranpatil) [reply src/components/Facility/FacilityConfigure.tsx]: done

File: src/components/Facility/LocationManagement.tsx (hunk 3 above)
Comment (Jacobjeevan) [reply src/components/Facility/LocationManagement.tsx]: There's a button component in components/ui/

Edit: I meant substitute the ButtonV2 with Button.

Comment (saikiranpatil) [reply src/components/Facility/LocationManagement.tsx]: @Jacobjeevan  buttonV2 has lot more attributes than just className. like href, tooltip, authorizefor, many others. we need to modify the shadcn button. how about i create a new issue to migrate from buttonV2 to shadcn button?

**Comment by saikiranpatil** [comment]
@ohcnetwork/care-fe-code-reviewers @Jacobjeevan buttonV2 has lot more attributes than just className. like href, tooltip, authorizefor, many others. we need to modify the shadcn button. how about i create a new issue to migrate from buttonV2 to shadcn button?

File: src/components/Facility/LocationManagement.tsx (hunk 3 above)
Comment (Jacobjeevan) [reply src/components/Facility/LocationManagement.tsx]: > @Jacobjeevan buttonV2 has lot more attributes than just className. like href, tooltip, authorizefor, many others. we need to modify the shadcn button. how about i create a new issue to migrate from buttonV2 to shadcn button?

No, don't modify Shadcn components. You can use AuthorizedButton instead, for example in this [PR](https://github.com/ohcnetwork/care_fe/pull/9377/files#diff-5196ad947a67663db5d6c94e15aaec530c9ac0d0c67bb7dde0a50d774a1ecafc); use corresponding logic with shadcn button for others (`Link` from raviger for hrefs).

File: src/components/Facility/FacilityConfigure.tsx [hunk 5]
```diff
@@ -1,166 +1,146 @@
...
       navigate(`/facility/${facilityId}`);
-    } else {
-      Notification.Error({
-        msg: error?.detail ?? "Something went wrong",
-      });
-    }
-    setIsLoading(false);
-  };
+    },
+  });
```
[the remark this answers is not in the document]
Comment (saikiranpatil) [reply src/components/Facility/FacilityConfigure.tsx:65]: useQuery errors are handled globally

**Comment by saikiranpatil** [comment]
@Jacobjeevan review the changes

**Comment by saikiranpatil** [comment]
@ohcnetwork/care-fe-code-reviewers @Jacobjeevan review the changes

**Comment by Jacobjeevan** [comment]
Closing for now, we will be reworking this anyways.

**Pull request closed without merging**